Add data type/schema field/schema

lidavidm opened the pull request:
Implement the API for data types and schemas.

lidavidm commented:
Note: we could simplify the data type definition greatly if we adopt something like Arrow Java/[cuDF](https://docs.rapids.ai/api/libcudf/stable/classcudf_1_1data__type). Type would be a POD containing type ID and fields for all parameterized types (decimal scale, fixed length, etc.). Field would contain a Type and a (possibly empty/invalid) list of nested fields (to handle things like Struct). This loses type safety but removes a layer of indirection/allocations.

lidavidm commented:
For here I've chosen something somewhat resembling Arrow C++. However I think Schema and Field (unlike Arrow) should not always be wrapped in smart pointers. Type is still in a smart pointer for type erasure.

lidavidm commented on src/iceberg/type_fwd.h:
```diff
@@ -0,0 +1,55 @@
+///
+/// This is not a complete data type by itself because some types are nested
+/// and/or parameterized.
+enum class TypeId {
```
This needs the Iceberg V3 types added as well.

wgtmac replied:
+1 for the 1st approach. Supporting v2 is already a big milestone.

lidavidm commented on src/iceberg/schema.h:
```diff
@@ -0,0 +1,60 @@
+/// A schema is a list of typed columns, along with a unique integer ID.  A
+/// Table may have different schemas over its lifetime due to schema
+/// evolution.
+class ICEBERG_EXPORT Schema : public StructType {
```
In iceberg-arrow we can add functions from iceberg::Schema to arrow::Schema.

wgtmac replied:
Do we need to add `KeyValueMetadata` to Schema and Field, though the Iceberg spec does not have it yet.

lidavidm replied:
I'd rather not add it if it's not in the spec.

lidavidm replied:
(Also if Iceberg is adding it, consider the issues that Arrow has: Arrow currently doesn't allow binary metadata values due to an oversight.)

Fokko reviewed (commented):
Looking good, left a few comments

wgtmac commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,120 @@
+  [[nodiscard]] std::span<Field> fields() const;
+
+  const Field& field(int i) const;
+  const Field& field(std::string_view name) const;
```
I believe this follows the Arrow style where trivial accessor functions are lower-cased.

lidavidm replied:
I could change this to be FieldByIndex/FieldByName perhaps

lidavidm replied:
Changed to GetFieldBy{Index,Name}

wgtmac commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,120 @@
+namespace iceberg {
+
+/// \brief Interface for a data type for a field.
+class ICEBERG_EXPORT Type {
```
Should we add default value or later?

lidavidm replied:
I don't see that in the Java API. We could add it as it becomes useful?

lidavidm commented:
I added Primitive/NestedType; I'll start filling out definitions and the rest of the types next, along with adding the visitors to iceberg-arrow.

gaborkaszab commented on src/iceberg/schema.h:
```diff
@@ -0,0 +1,60 @@
+
+#pragma once
+
+/// \file iceberg/schema.h
```
I'm not familiar with this style of commenting. Could you please explain me why including the file name into a comment needed here?

lidavidm replied:
I'm not done with it yet, this is for Doxygen which will generate a page to document the header.

gaborkaszab commented on src/iceberg/schema.h:
```diff
@@ -0,0 +1,60 @@
+  ///
+  /// Schemas are identified by a unique ID for the purposes of schema
+  /// evolution.
+  [[nodiscard]] int32_t schema_id() const;
```
The functions within this class may not follow the same naming convention: See schema_id vs ToString, Equals
I found the same mismatch below too, so I'm wondering if it's me missing something here.

lidavidm replied:
Google style has trivial accessors using snake_case.

gaborkaszab commented on src/iceberg/schema.h:
```diff
@@ -0,0 +1,60 @@
+  [[nodiscard]] std::string ToString() const;
+
+  /// \brief Compare two schemas for equality.
+  [[nodiscard]] bool Equals(const Schema& other) const;
```
If we have operators for equality, can't we make the Equals() function private? I don't see the reason to expose both.

lidavidm replied:
It may be useful to add overloads later (e.g. if we want to ignore metadata or IDs)

gaborkaszab replied:
Then we can extend this later. I believe in adding what is required ATM, and if there is a need later, we can still extend.
Also, we are defining a public interface here, should keep it as simple as possible. Exposing the Equals method seems a redundancy for me ATM.

gaborkaszab commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,149 @@
+  /// \param[in] name The field name.
+  /// \param[in] type The field type.
+  /// \param[in] optional Whether values of this field are required or nullable.
+  Field(int32_t field_id, std::string name, std::shared_ptr<Type> type, bool optional);
```
> std::shared_ptr<Type> type
Does this assume that for each type there is going to be an instance and the Fields will share these instances? Would it be too much overhead to keep a copy of a Type for each Field?

gaborkaszab replied:
Well, I think it's been too long for me since I found myself writing Java code and not C++ on a daily basis :) After taking another look, sharep_ptr<Type> seems fine here. Sorry :)

gaborkaszab commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,149 @@
+namespace iceberg {
+
+/// \brief A type combined with a name.
+class ICEBERG_EXPORT Field {
```
I think similarly to Schema, I'd put Field into a separate file. Now this is in type.h but Field isn't really a type.

lidavidm replied:
There's a bit of a circular dependency between Field (which wants Type) and StructType (which wants Field). Unless (again) you see the alternate design in the first two comments of this PR which puts child Fields always on the Field (this means that Field has a potentially redundant...field, but avoids nesting of types - I personally slightly prefer this but it's conceptually dependent on whether you model the type itself as nested or the field as the nested object)

gaborkaszab replied:
Sure, will take a look a bit later.
Generally asking, can't we solve the circular dependency with forward declarations and using pointers/references?

lidavidm replied:
We can if we continue to type-erase Type, but not if we want it to not be behind a smart pointer.

gaborkaszab replied:
Putting them behind smart pointers makes sense. See my other comment. And the this could go into a separate file.
Meanwhile I'm sketching some code for the PartitionSpec and PartitionField and I'm wondering if this one here is intuitive to be called Field, or could be better called SchemaField (considering there is going to be a PartitionField too).

lidavidm replied:
I'll rename

wgtmac replied:
Do you suppose `PartitionField` to inherit `SchemaField` or it is totally a separate thing? @gaborkaszab

gaborkaszab reviewed (commented):
Great to see progress on this area, @lidavidm ! I went through for my own education and left some comments or questions. Let me know what you think!

gaborkaszab commented on src/iceberg/schema.h:
```diff
@@ -0,0 +1,60 @@
+/// evolution.
+class ICEBERG_EXPORT Schema : public StructType {
+ public:
+  Schema(int32_t schema_id, std::vector<Field> fields);
```
One thing I saw in the Java implementation, is that there is a name to field ID mapping there, and there are accessors getting Field(s) by ID or by name. I think that would be useful here too.

gaborkaszab commented on src/iceberg/schema.h:
```diff
@@ -0,0 +1,60 @@
+/// A schema is a list of typed columns, along with a unique integer ID.  A
+/// Table may have different schemas over its lifetime due to schema
+/// evolution.
+class ICEBERG_EXPORT Schema : public StructType {
```
> : public StructType
I don't think that Schema should derive from StructType. I get that it will hold a list of Fields such as StructType, but still, Schema in general is not a Type, would be pretty unnatural to derive from it.

lidavidm replied:
(Note that various type systems do actually just represent a schema as a struct type, including nanoarrow/Arrow C Data Interface. At the very least they are isomorphic and you can see methods that convert between them often. That said I don't feel strongly about this.)

wgtmac replied:
I'm also inclined to inherit Schema from StructType. In practice this inheritance avoid an unnecessary conversion from Schema to StructType to implement type visitors.

lidavidm replied:
That would also work, it would just add the conversion, and then duplicating all the methods. No big deal either way to me at least.

(I suppose the OOP approach is to add a StructLike interface that both Struct and Schema implement.)

lidavidm commented:
Just to make sure @gaborkaszab @wgtmac: are we ok with the Arrow-style type representation here? (Types are represented by a class hierarchy, erased behind smart pointers; nested types store the fields in the type object, not the field object)

The alternatives are:
- cuDF style: type objects are minimal, just a type ID, and there is no hierarchy, just the base `DataType` class which is not type-erased and therefore there is no need for a smart pointer. As a trade-off, nested fields have to be extracted from the field and the base DataType has to have fields for all possible parameterized types
- arrow-java style: there is still a hierarchy of type objects, but child fields are stored on the Field, not the DataType. Avoids a conceptual dependency cycle between DataType and Field
- variant style: like arrow-java, but instead of type erasure and a type hierarchy (or conversely, like cuDF but avoids redundant fields), we have a `std::variant` with all possibilities. Avoids boxing but the object is larger (in e.g. a vector)

lidavidm commented:
Arrow-style does let you do a bunch of compile-time metaprogramming (e.g. see arrow::TypeTraits)

mapleFU commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,149 @@
+  /// \brief Is this a primitive type (may not have child fields)?
+  [[nodiscard]] virtual bool is_primitive() const = 0;
+
+  /// \brief Is this a nested type (may have child fields)?
```
What raise my interests is that:
1. Does iceberg has a "void" ( or null ) bottom type?
2. Does iceberg allowed to define an empty structure?

lidavidm replied:
There is `unknown`. 

I don't think the spec says anything either way about an empty struct. I didn't check any implementations. If allowed, you could think of it as a `void` type.

Fokko replied:
The `UnknownType` will be part of the V3 specification.

An empty struct is pretty common (for example, an unpartitioned table has an empty struct as a partition).

gaborkaszab commented:
> Just to make sure @gaborkaszab @wgtmac: are we ok with the Arrow-style type representation here? (Types are represented by a class hierarchy, erased behind smart pointers; nested types store the fields in the type object, not the field object)

If I get it right, Arrow-style mes sense for me too.

lidavidm commented:
I've implemented the Type subclasses. I've also added a trait for ToString/std::format formatting (might as well use the new C++ features if we're requiring C++20). Although it appears clang is not happy about what I did.

I've elected to make getters fallible (e.g. GetFieldById) via `optional<reference_wrapper<T>>` which looks a bit messy but should be safer than returning `T const&` and throwing, or returning `T const*` and admitting `nullptr`.

lidavidm commented:
Ok, hopefully that fixes macOS. It appears C++20 support on the version of clang shipped by macOS is still flaky:

https://godbolt.org/z/W9rTexc13

Maybe we want to avoid C++20 (sadly) or maybe we need to encourage people to get a newer clang on macOS. (It appears to be clang 15, whereas the latest is clang 19.) Or maybe it's just an issue with the version the CI environment has.

lidavidm commented:
I'll add some unit tests as long as the API looks good to everyone.

wgtmac commented on src/iceberg/schema_field.h:
```diff
@@ -0,0 +1,87 @@
+                                  std::shared_ptr<Type> type);
+
+  /// \brief Get the field ID.
+  [[nodiscard]] int32_t field_id() const;
```
Is it an overkill to add `[[nodiscard]]` to these accessors?

wgtmac replied:
Right, it is just personal flavor. I have commented at https://github.com/apache/iceberg-cpp/pull/32#issuecomment-2611657586

wgtmac commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,397 @@
+  GetFieldById(int32_t field_id) const = 0;
+  /// \brief Get a field by index.
+  [[nodiscard]] virtual std::optional<std::reference_wrapper<const SchemaField>>
+  GetFieldByIndex(int i) const = 0;
```
```suggestion
  GetFieldByIndex(int32_t i) const = 0;
```

I'd prefer explicit width

lidavidm replied:
Do we want signed or unsigned?

wgtmac replied:
Perhaps int32_t as my suggested change.

wgtmac commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,397 @@
+/// \brief A data type representing a fixed-precision decimal.
+class ICEBERG_EXPORT DecimalType : public PrimitiveType {
+ public:
+  constexpr static const int32_t kMaxPrecision = 38;
```
Is it worth adding enforcement of constant style to `.clang-tidy`? Now it seems that we prefer `kMaxPrecision` over `MAX_PRECISION`

lidavidm replied:
We should add it. Maybe this isn't in line with the style guide

lidavidm replied:
This is correct: https://google.github.io/styleguide/cppguide.html#Constant_Names

wgtmac commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,397 @@
+
+/// \brief A data type representing a timestamp in microseconds in a
+///   particular timezone.
+class ICEBERG_EXPORT TimestampTzType : public PrimitiveType {
```
Should we add an abstract base class for all timestamp types? We will need `timestamp_ns` and `timestamptz_ns` for v3. The only differences are `timezone` and `precision`.

lidavidm replied:
Are timezone types parameterized? The Java implementation is weird and only has `adjustToUTC` (so I guess there are only naive and UTC timestamps in Iceberg?): https://github.com/apache/iceberg/blob/ab9e05e7369c55a95632b856729a05ea876cf788/api/src/main/java/org/apache/iceberg/types/Types.java#L211

wgtmac replied:
Right, Parquet and Iceberg only care about whether this is normalized to UTC.

lidavidm replied:
Hmm, I can add a base class then.

wgtmac reviewed (commented):
The interface looks good to me. Thanks!

lidavidm commented:
I added a bunch of unit tests + fixed some copy-paste errors I discovered in the process.

mapleFU commented on src/iceberg/type.cc:
```diff
@@ -0,0 +1,314 @@
+  if (precision < 0 || precision > kMaxPrecision) {
+    throw std::runtime_error(
+        std::format("DecimalType: precision must be in [0, 38], was {}", precision));
+  }
```
Now that we've checked the precision, should we also check the scale here?

lidavidm replied:
Java actually doesn't validate the scale, and the spec is silent on this part. (Also, Java actually allows negative precision, though I don't think that makes any sense?)

mapleFU replied:
From https://github.com/apache/iceberg/issues/7420 I think setting it or not is all possible

zhjwpku replied:
Blue give some examples which scale can be greater than precision and smaller than 0[1], though I don't think it makes sense, I agree we should keep this as is.

[1] https://github.com/apache/iceberg/issues/7420#issuecomment-1526522715

lidavidm replied:
Can we fix this in the spec? The spec itself says nothing about these types which is a little frustrating since I feel like the Java implementation is the actual de-facto spec

lidavidm replied:
Oh wow, the issue asking for spec clarification was closed. Alright then, I will leave this as-is

mapleFU commented on src/iceberg/type.cc:
```diff
@@ -0,0 +1,314 @@
+  }
+}
+
+int32_t FixedType::length() const { return length_; }
```
an unimportant issue: should we just use uint32_t?

lidavidm replied:
Java uses int, so I'd rather be consistent

mapleFU replied:
Seems that Java not has unsigned int? But both part lgtm

lidavidm replied:
Java does not have unsigned types but you can manipulate signed types as if they were unsigned when necessary. I'll close this since there's another thread about this now...

mapleFU commented on src/iceberg/type.cc:
```diff
@@ -0,0 +1,314 @@
+                      field.field_id(), it->second, index));
+    }
+
+    index++;
```
nit: ++index;

lidavidm replied:
Fixed

mapleFU commented on src/iceberg/type.cc:
```diff
@@ -0,0 +1,314 @@
+}
+std::optional<std::reference_wrapper<const SchemaField>> StructType::GetFieldByName(
+    std::string_view name) const {
+  // TODO: what is the right behavior if there are duplicate names? (Are
```
I found that the java and the rust implement has a lazy initialization and locked hash map in get field by name, but I don't understand the reason...

lidavidm replied:
The reason for the laziness, or the reason for the map? (I could change this to be lazy but tbh that introduces quite a bit of complication...)

mapleFU replied:
I prefer just do like this until someone need optimize this!

mapleFU reviewed (commented):
Great work!

zhjwpku commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,411 @@
+  bool Equals(const Type& other) const override;
+
+ private:
+  int32_t length_;
```
iceberg-rust has the length of `u64`[1], `int32_t` can represent about 2 gigabytes, maybe we should change to `int64_t` to be compatible with other language implementation?

[1] https://github.com/apache/iceberg-rust/blob/main/crates/iceberg/src/spec/datatypes.rs#L241

lidavidm replied:
Java uses a 32 bit int. https://github.com/apache/iceberg/blob/d693f83b6ad6e306550d117ff0bf0014356500b2/api/src/main/java/org/apache/iceberg/types/Types.java#L360

Are the Iceberg maintainers interested in making the spec actually lay out these details? I've mostly copied Java since the spec itself is extremely vague.

Fokko replied:
The int32 is fine. No one should go beyond 2GB values for a single value.

zhjwpku commented on src/iceberg/type_fwd.h:
```diff
@@ -0,0 +1,85 @@
+  kTime,
+  kTimestamp,
+  kTimestampTz,
+  kBinary,
```
nit: list these the same order as spec.

lidavidm replied:
Is the spec actually canonical? It seems the implementations all disagree...

lidavidm replied:
And the spec doesn't actually have any single listing of all the types (since nested types are split off), so I am electing to put the nested types first to be consistent with the spec (in as far as it defines any order)

lidavidm replied:
Updated the order

lidavidm commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,411 @@
+  [[nodiscard]] virtual bool is_primitive() const = 0;
+
+  /// \brief Is this a nested type (may have child fields)?
+  [[nodiscard]] virtual bool is_nested() const = 0;
```
To me it's clearer to be able to say `is_nested` or `is_primitive` and they can just be implemented in terms of each other.

zhjwpku replied:
Ok, let's keep both.

lidavidm commented on src/iceberg/type_fwd.h:
```diff
@@ -0,0 +1,85 @@
+class Schema;
+class SchemaField;
+class StringType;
+class StructType;
```
Fixed

lidavidm commented on src/iceberg/schema.h:
```diff
@@ -0,0 +1,64 @@
+
+  /// \brief Get the schema ID.
+  ///
+  /// Schemas are identified by a unique ID for the purposes of schema
```
Changed to "A schema..."

lidavidm commented on src/iceberg/type.cc:
```diff
@@ -0,0 +1,314 @@
+  return other.type_id() == TypeId::kBoolean;
+}
+
+TypeId Int32Type::type_id() const { return TypeId::kInt32; }
```
Updated

lidavidm commented on src/iceberg/type.cc:
```diff
@@ -0,0 +1,314 @@
+  return precision_ == decimal.precision_ && scale_ == decimal.scale_;
+}
+
+TypeId TimeType::type_id() const { return TypeId::kTime; }
```
Reordered headers/impls

lidavidm commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,411 @@
+  [[nodiscard]] virtual bool Equals(const Type& other) const = 0;
+};
+
+/// \brief A data type that may not have child fields.
```
Updated

lidavidm commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,411 @@
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a 32-bit (single precision) float.
```
Updated

wgtmac commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,414 @@
+};
+
+/// \brief A data type that has child fields.
+class ICEBERG_EXPORT NestedType : public Type {
```
Do we need a function for getting number of fields?

wgtmac replied:
That's enough I think.

wgtmac commented on src/iceberg/type.h:
```diff
@@ -0,0 +1,414 @@
+  /// \brief Get a field by index.
+  [[nodiscard]] virtual std::optional<std::reference_wrapper<const SchemaField>>
+  GetFieldByIndex(int32_t index) const = 0;
+  /// \brief Get a field by name.
```
Is this case sensitive?

lidavidm replied:
Updated

wgtmac commented on src/iceberg/type.cc:
```diff
@@ -0,0 +1,312 @@
+  for (const auto& field : fields_) {
+    auto [it, inserted] = field_id_to_index_.try_emplace(field.field_id(), index);
+    if (!inserted) {
+      throw std::runtime_error(
```
Should we define `IcebergException` or even a more specific exception to replace it?

lidavidm replied:
Are we using exceptions in general? I am only using an exception here because this is a constructor. We could also hide the constructor, or change to assertions.

wgtmac replied:
No, my point is that a more specific exception is better than the `std::runtime_error`. 

BTW, should we build `field_id_to_index_` lazily when the particular getter is called for the 1st time?

lidavidm replied:
That's discussed above, but it complicates the implementation a lot

wgtmac commented on src/iceberg/type.cc:
```diff
@@ -0,0 +1,312 @@
+
+TypeId StructType::type_id() const { return TypeId::kStruct; }
+std::string StructType::ToString() const {
+  std::string repr = "struct<\n";
```
Is there a specific reason to use a separate line for each sub-field? An alternative is to use the Hive-style where all fields are concatenated in a single line.

lidavidm replied:
That gets messy IMO once there are more than a couple fields

wgtmac replied:
Do we want to support restoring a Schema object by deserializing the formatted string?

lidavidm replied:
I'd rather add JSON serializers than try to parse the repr

wgtmac commented on src/iceberg/type.cc:
```diff
@@ -0,0 +1,312 @@
+}
+std::optional<std::reference_wrapper<const SchemaField>> StructType::GetFieldByName(
+    std::string_view name) const {
+  // TODO: what is the right behavior if there are duplicate names? (Are
```
It seems that the spec is unclear about this so the behavior is undefined and implementation-specific.

lidavidm replied:
There seem to be a lot of spec issues...

Fokko replied:
Please consider the Java implementation as the reference one. There we raise a `ValidationException` when we see a duplicate name:

https://github.com/apache/iceberg/blob/507e2a9e39f93356bdc0dae1521bbdc46629681c/api/src/main/java/org/apache/iceberg/types/IndexByName.java#L199-L205

> There seem to be a lot of spec issues...

@lidavidm Please elaborate

lidavidm replied:
Is this in the written spec? Or is the canonical-ness of the Java implementation documented somewhere? Just so I know what to reference

lidavidm replied:
Also see above e.g. https://github.com/apache/iceberg-cpp/pull/31#discussion_r1929559679
https://github.com/apache/iceberg-cpp/pull/31#discussion_r1928875579

possibly I overlooked something

Fokko replied:
The spec describes the table format itself, for the behavioral things, we tend to check the Java implementation (considered the reference implementation). If we would need to formalize all this as well, then the spec would become unreadable.

lidavidm replied:
IMO, at least something like "duplicate field names are not allowed" should be part of the spec, no? I wouldn't really qualify that as "behavioral" any more than "decimal precision is limited to 38" is behavioral

Fokko replied:
Since Iceberg comes from the database world, not having duplicate names is more or less implied to me, but feel free to raise a PR to add this to the spec itself.

I was surprised to see that this is allowed:

```python
>>> import pyarrow as pa
>>> pa.table([
...     pa.array(["foo", "bar"]),
...     pa.array(["Parrot", "Dog"])
... ], names=['animals', 'animals'])
pyarrow.Table
animals: string
animals: string
----
animals: [["foo","bar"]]
animals: [["Parrot","Dog"]]
```

wgtmac replied:
> I was surprised to see that this is allowed:

I think Arrow is a different. It is totally valid to project the same column more than once in an Arrow Table or RecordBatch. For example, returning query result via Arrow for `select animals, animals from foo.bar`;

lidavidm commented on src/iceberg/util/formatter.h:
```diff
@@ -0,0 +1,43 @@
+
+#pragma once
+
+/// \file iceberg/util/formatter.h
```
No, you do need this file externally if you yourself wish to use std::format.

lidavidm commented on src/iceberg/schema.cc:
```diff
@@ -0,0 +1,47 @@
+
+#include "iceberg/schema.h"
+
+#include <format>
```
Isn't it good practice to IWYU?

wgtmac replied:
I'm not that familiar with IWYU yet...

lidavidm replied:
We use std::format in the .cc file so we should include format, even if other headers happen to include it already

wgtmac reviewed (approved):
+1

lidavidm commented:
Is anything else required here?

zhjwpku commented:
> Is anything else required here?

+1 to merge this PR, thanks for working on this.

wgtmac commented:
@Fokko @Xuanwo I think this is ready to merge. cc @gaborkaszab in case there is any comment.

gaborkaszab commented:
> @Fokko @Xuanwo I think this is ready to merge. cc @gaborkaszab in case there is any comment.

I got a bit busy with other stuff, so wouldn't block this PR for me to find time here. Thanks for these changes!

Fokko commented on src/iceberg/schema_field.cc:
```diff
@@ -0,0 +1,64 @@
+    : field_id_(field_id),
+      name_(std::move(name)),
+      type_(std::move(type)),
+      optional_(optional) {}
```
It is because the spec serializes it as `required`: https://iceberg.apache.org/spec/#schemas This way it is more straightforward to serialize it using the library that we use (Pydantic). See the original issue: https://github.com/apache/iceberg/issues/4985. In PyIceberg we have a property called `optional` as well, so you can choose.

Fokko commented on src/iceberg/schema_field.cc:
```diff
@@ -0,0 +1,64 @@
+bool SchemaField::optional() const { return optional_; }
+
+std::string SchemaField::ToString() const {
+  return std::format("{} ({}): {}{}", name_, field_id_, *type_,
+                     optional_ ? "" : " (required)");
```
Maybe it is just me, but I'm not a big fan of the implied `(optional)`

Fokko commented on src/iceberg/type.cc:
```diff
@@ -0,0 +1,312 @@
+  }
+  return fields_[index];
+}
+std::optional<std::reference_wrapper<const SchemaField>> StructType::GetFieldByName(
```
We also want to distinguish upper- and lower-case

lidavidm replied:
The docstring states this is case-sensitive already. Should we add something like Java's `StructType#caseInsensitiveField`?

Fokko reviewed (approved):
Looks like there is consensus, let's move this forward 🚀

lidavidm commented:
Thanks @Fokko I'll follow up to correct those things
